keep _parent field while updating child type mapping

PnPie opened the pull request:
Be able to update child type mapping without specifying it's `_parent` field, because it should be considered as unchanged if it is not explicitly specified.
In this case, the merged mapper's `parentType` field should be `null`, so we can just merge it to keep the original one instead of throwing an exception of "trying to change it to 'null' "

Close #23381

PnPie committed:
Keep _parent field mapping while updating child mapping (#23381)

be able to update child type mapping without specifying _parent
field instead of trying to update it to "null" as previously.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

clintongormley commented:
@colings86 please could somebody review this

martijnvg commented on core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperMergeTests.java:
```diff
@@ -235,4 +235,33 @@ public void testDoNotRepeatOriginalMapping() throws IOException {
+            .startObject("_parent").field("type", "new_parent").endObject()
+            .endObject().endObject().string();
+        DocumentMapper modParentMapper = parser.parse("child", new CompressedXContent(modParentMapping));
+        try {
```
Replace this try-catch block with an `expectThrows(...)`  call?

```java
Exception e = expectThrows(IllegalArgumentException.class, () ->  initMapper.merge(modParentMapper.mapping(), false))
assertThat(e.getMessage(), containsString("The _parent field's type option can't be changed:"));
```

martijnvg commented on core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperMergeTests.java:
```diff
@@ -235,4 +235,33 @@ public void testDoNotRepeatOriginalMapping() throws IOException {
+            .endObject().endObject().endObject().string();
+        DocumentMapper updatedMapper = parser.parse("child", new CompressedXContent(updatedMapping));
+        DocumentMapper mergedMapper = initMapper.merge(updatedMapper.mapping(), false);
+
```
Maybe also add a merge  with current parent field and a new field?

```java
String updatedMapping = XContentFactory.jsonBuilder().startObject().startObject("child")
   .startObject("_parent").field("type", "parent").endObject()
   .startObject("properties")
   .startObject("field2").field("type", "text").endObject()
   .endObject().endObject().endObject().string();
 DocumentMapper updatedMapper = parser.parse("child", new CompressedXContent(updatedMapping));
 DocumentMapper mergedMapper = initMapper.merge(updatedMapper.mapping(), false);
```

martijnvg reviewed (commented):
Thanks @PnPie! This looks good. I left two comments regarding the test.

PnPie committed:
Fix _parent field's name changing while merging

The parent field should not be updatable, but the it's fieldType is changed and the new one is applied. So this patch make it keep the original one while merging.

PnPie committed:
Merge github.com:elastic/elasticsearch

PnPie commented:
Thanks for your comments @martijnvg, I added a merge case with current parent field while I found a new problem: I cannot find the `_parent` field with this assertion didn't pass
`assertThat(mergedMapper.mappers().getMapper("_parent"), notNullValue());`

And then I found that the parent field name of a document is in format: `_parent#[parent type]`, so in the test it should be `_parent#parent`, if a new mapper without parent field, it's name is just `_parent`, and when we do the merge we firstly call the `super.doMerge` method of `ParentFieldMapper` and it always replaces the `fieldType` with the new one's, so in consequence the parent field name is changed to the new one's.
https://github.com/PnPie/elasticsearch/blob/346889a93bb18cda93736a8076aef198b8f86fe1/core/src/main/java/org/elasticsearch/index/mapper/ParentFieldMapper.java#L298
So previously after this we can find the `_parent` field with
`assertThat(mergedMapper.mappers().getMapper("_parent"), notNullValue());`
but in fact the parent field shouldn't be changed and it should be `_parent#parent`, so I modified in `ParentFieldMapper`'s `doMerge` method and make it keep the original parent field name.

And I also add a test for trying to add parent field to an exiting mapper without parent.

martijnvg commented:
Cool, looks good @PnPie. I'll merge it in soon and backport to 5.x branch.
